[WIP] Segregated Witness

jonasschnelli opened the pull request:
This PR should allow people to look at the current implementation.
[work in progress]

sipa committed:
Add segregated witness transaction serialization

Contains refactorings by Eric Lombrozo.
Contains fixup by Nicolas Dorier.

sipa committed:
Witness commitment validation

Includes a fix by Suhas Daftuar

sipa committed:
Script validation logic for witnesses

sipa committed:
Introduce and preferentially peer with NODE_WITNESS service bit

Service bit logic by Nicolas Dorier.

sipa committed:
Enable SCRIPT_VERIFY_WITNESS for mempool transactions

sipa committed:
Activate script consensus rules through BIP9

sipa committed:
Signing tests

sipa committed:
Add signature version 1 with updated sighash

Includes simplifications by Eric Lombrozo.

sipa committed:
Only download blocks from witness peers after fork

sipa committed:
Add witness address RPCs (using P2SH)

Includes support for pushkeyhash wit v0 by Alex Morcos.

sipa committed:
Create segnet4

morcos committed:
Add command line options to loosen mempool acceptance rules

morcos committed:
Add rpc test for segwit

Amended by Pieter Wuille to use multisig 1-of-1 for P2WSH tests, and BIP9
based switchover logic.

sipa committed:
Implement block size/sigop cost rules, limits, and GBT support

Includes changes by Suhas Daftuar and Luke-jr.

CodeShark committed:
Removed ppszTypeName from protocol.cpp

theuni committed:
qt: Work (don't crash) with -segnet

CodeShark committed:
getdata enum issue fix

sdaftuar committed:
Increase MAX_PROTOCOL_MESSAGE_LENGTH

Witness blocks can be greater than 2MiB, but cannot be validly greater
than 4MB.

sipa committed:
Add segnet seed nodes

jl2012 committed:
Return witness data

Includes RPC field name changes by Luke-jr.

sadshoe1 commented:
NACK.

SegWit splits bitcoin blocks into two economically distinct limited resources. There has been no economic analysis to justify this.

SegWit eliminates Bitcoin security by depriving nodes of the ability to validate signatures.

The Bitcoin community REJECTS the conversion of the Bitcoin network into a settlement network.

There is no consensus for this highly controversial change and it does not even pass the tests.

NicolasDorier commented:
@sadshoe1 it seems that you are very new to github (today) so you may not understand what a PR discussion is.
A PR discussion is for talking about technical problems seen in the code implementing this idea. If you want to debate the idea itself, on ground other than technical, maybe reddit is more appropriate.

afk11 committed:
bitcoinconsensus: add method that accepts amount, and return error when verify_script receives VERIFY_WITNESS flag

script_tests: always test bitcoinconsensus_verify_script_with_amount if VERIFY_WITNESS isn't set

Rename internal method + make it static

trim bitcoinconsensus_ prefix

Add SERIALIZE_TRANSACTION_WITNESS flag

laanwj commented:
I've deleted the post by @sadshoe. Please limit the discussion here to the code only.

rebroad commented on src/chainparams.cpp:
```diff
@@ -77,6 +77,7 @@ class CMainParams : public CChainParams {
         consensus.BIP34Height = 227931;
         consensus.BIP34Hash = uint256S("0x000000000000024b89b42a942fe0d9fea3bb44ab7bd1b19115dd6a759c0808b8");
         consensus.powLimit = uint256S("00000000ffffffffffffffffffffffffffffffffffffffffffffffffffffffff");
+        consensus.SegWitHeight = 2000000000;
```
what is this number?

sipa replied:
"some long time in the future" (as we don't have a planned activation date for segwit on mainnet yet). actual segwit will be deployed using versionbits, so this will go away

rebroad commented on src/chainparamsseeds.h:
```diff
@@ -889,6 +889,13 @@ static SeedSpec6 pnSeed6_main[] = {
     {{0xfd,0x87,0xd8,0x7e,0xeb,0x43,0xcf,0x7b,0x5e,0x3a,0x53,0x21,0x5b,0x62,0xe3,0x7a}, 8333}
 };
 
+static SeedSpec6 pnSeed6_seg[] = {
```
would be nice if the code has more comments that referred to documentation that explains what some of these numbers are

laanwj committed:
test: Re-introduce JSON pretty printing in test builder

laanwj committed:
test: Script_error checking in script_invalid tests

Check the returned script_error. Add expected script_error
for generated as well as custom tests.

The specific error is not part of consensus, however
it could avoid unclear reporting issues such as #6862 in the future.

Fixes #7513.

sipa committed:
Add rewind logic to deal with post-fork software updates

sipa committed:
Only store and connect to NODE_NETWORK nodes

sipa committed:
Keep addrman's nService bits consistent with outbound observations

sipa committed:
Verify that outbound connections have expected services

Commit:
[qa] wallet: Wait for reindex to catch up

sipa committed:
--- [SEGWIT] begin: segnet ---

sipa committed:
--- [SEGWIT] begin: P2P/node/consensus ---

sipa committed:
--- [SEGWIT] begin: wallet ---

sipa committed:
--- [SEGWIT] begin: tests ---

sipa committed:
Observe input amounts: verification

sipa committed:
Witness script signing

NicolasDorier committed:
Add transaction tests for segwit

...with the four types of segwit payment, as well as all sighash combinaisons.

sipa committed:
Add segwit support to script_tests

sipa committed:
Autogeneration support for witness in script_tests

sipa committed:
Fix some misconstructed tests

They claimed to be testing P2SH scripts with non-push scriptSigs, but
1) they were not enabling P2SH
2) they have push-only scriptSigs

Fix this, and add a few more related cases.

sipa committed:
Fix formatting of NOPs for generated script tests

sipa committed:
Fix JSON pretty printing in script_tests

sipa committed:
Make script_error a mandatory 4th field for script_tests

sipa committed:
Get rid of expect in script_tests as it's implied by scripterror

sipa committed:
Merge script_valid and script_invalid tests

sdaftuar committed:
Update p2p test framework with segwit support

mininode now supports witness transactions/blocks, blocktools
has a helper for adding witness commitments to blocks, and script
has a function to calculate hashes for signature under sigversion
1, used by segwit.

sdaftuar committed:
P2P test for segwit

NicolasDorier committed:
signrawtransaction can sign P2WSH

sdaftuar committed:
Use txid as key in mapAlreadyAskedFor

Previously we used the CInv that would be sent to the peer announcing the
transaction as the key, but using the txid instead allows us to decouple the
p2p layer from the application logic (which relies on this map to avoid
duplicate tx requests).

sipa committed:
Change mapRelay to store CTransactions

sipa committed:
Don't check the genesis block PoW

Inspired by Jorge Timon's "Testchain: Don't check the genesis block", but
restricted to just PoW check.

Code by Matt Corallo, originally submitted as a patch to segwit.

jonasschnelli commented:
Closing in favor of #7910
